### PR TITLE
[WebProfilerBundle] remove definitions when they are not usable

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Comment by stof** [comment]
Is anything usable in WebProfilerBundle if you don't have Twig ? I don't think so.

**Comment by stof** [comment]
WebProfilerBundle should have a requirement on TwigBundle, not just on Twig. It would solve the issue cleanly in a Flex world when people install things separately.

**Comment by xabbuh** [comment]
Probably not, but this at least prevents the `cache:clear` call from breaking.

**Comment by stof** [comment]
I think the reason to do it this way was to enable usage of the WebProfilerBundle code in Silex without bringing FrameworkBundle. But TwigBundle no longer depend on FrameworkBundle. The only dependency that TwigBundle has and WebProfilerBundle does not have is `symfony/config` (which won't hurt in the context of Silex, and is necessary in the context of Symfony anyway)

**Comment by xabbuh** [comment]
@fabpot What do you think? Is it okay for you to add the TwigBundle requirement?

**Comment by fabpot** [comment]
The profiler does not need the bundle, just plain Twig as demonstrated in Silex. I would not add the bundle as a hard requirement. That being said, I can see that with Flex, more and more people have problems with this (even if using `composer req profiler` does the right thing).

Not sure what to do here.

**Comment by xabbuh** [comment]
In that case, just removing the invalid definitions as done in this PR should improve DX as clearing the cache afterwards shouldn't fail anymore.

**Comment by stof** [comment]
Well, it needs the bundle to have a working twig service (which is what the bundle is doing).

Regarding the usage in Silex, TwigBundle does depend on FrameworkBundle anyway, so having it as a dependency would not bring all of symfony anymore (it has the same dependencies than WebProfilerBundle)

File: src/Symfony/Bundle/WebProfilerBundle/DependencyInjection/Compiler/RemoveInvalidDefinitionsPass.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+class RemoveInvalidDefinitionsPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->has('twig')) {
+            $container->removeDefinition('web_profiler.controller.exception');
+            $container->removeDefinition('web_profiler.controller.profiler');
+            $container->removeDefinition('web_profiler.controller.router');
+            $container->removeDefinition('web_profiler.debug_toolbar');
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/DependencyInjection/Compiler/RemoveInvalidDefinitionsPass.php]: Removing all these makes the bundle useless, as the **only** goal of the bundle is to provide the UI to access the profiler data.

Comment (xabbuh) [reply src/Symfony/Bundle/WebProfilerBundle/DependencyInjection/Compiler/RemoveInvalidDefinitionsPass.php]: Sure, but at least clearing the cache does not fail now. I don't care which solution we choose, but anything is better DX-wise than the current behaviour.

**Comment by nicolas-grekas** [comment]
A way to solve this would be to allow flex to replace a package by a pack that would be auto-unpacked:
`composer require symfony/web-profiler-bundle` => flex replaces by profiler-pack (or another smaller pack that has the bare minimum to make this work in a Flex managed context) + auto-unpacks the pack.

This PR looks like a workaround/hack for a missing behavior in Flex.

**Comment by xabbuh** [comment]
To be fair, this is not limited to Flex but you would observe the same behaviour when installing the package manually using Composer without Flex.

**Comment by nicolas-grekas** [comment]
> same behaviour when installing the package manually

but then it's your responsibility to fix your dependencies

**Comment by nicolas-grekas** [comment]
So, shall we do this untill https://github.com/symfony/flex/issues/360 is implemented? WDYT?

**Comment by fabpot** [comment]
First, WebProfilerBundle never had a dependency on FrameworkBundle, so that's not the reason (and TwigBundle never had a dependency on FrameworkBundle either).

But thinking about this one again, I've changed my mind and I think that we can add TwigBundle as a dependency. Let's do it for 4.2 as Silex won't be maintained anymore when it will be released.

See #27417

**Comment by fabpot** [comment]
Don't you think that throwing an exception would be better?

**Pull request closed without merging**